File: FirebaseDatabase/Sources/Core/Utilities/FPath.m
/*
 * Copyright 2017 Google
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#import "FirebaseDatabase/Sources/Core/Utilities/FPath.h"

#import "FirebaseDatabase/Sources/Utilities/FUtilities.h"

@interface FPath ()

@property(nonatomic, readwrite, assign) NSInteger pieceNum;
@property(nonatomic, strong) NSArray *pieces;

@end

@implementation FPath

#pragma mark -
#pragma mark Initializers

+ (FPath *)relativePathFrom:(FPath *)outer to:(FPath *)inner {
    NSString *outerFront = [outer getFront];
    NSString *innerFront = [inner getFront];
    if (outerFront == nil) {
        return inner;
    } else if ([outerFront isEqualToString:innerFront]) {
        return [self relativePathFrom:[outer popFront] to:[inner popFront]];
    } else {
        @throw [[NSException alloc]
            initWithName:@"FirebaseDatabaseInternalError"
                  reason:[NSString
                             stringWithFormat:
                                 @"innerPath (%@) is not within outerPath (%@)",
                                 inner, outer]
                userInfo:nil];
    }
}

+ (FPath *)pathWithString:(NSString *)string {
    return [[FPath alloc] initWith:string];
}

- (id)initWith:(NSString *)path {
    self = [super init];
    if (self) {
        NSArray *pathPieces = [path componentsSeparatedByString:@"/"];
        NSMutableArray *newPieces = [[NSMutableArray alloc] init];
        for (NSInteger i = 0; i < pathPieces.count; i++) {
            NSString *piece = [pathPieces objectAtIndex:i];
            if (piece.length > 0) {
                [newPieces addObject:piece];
            }
        }

        self.pieces = newPieces;
        self.pieceNum = 0;
    }
    return self;
}

- (id)initWithPieces:(NSArray *)somePieces andPieceNum:(NSInteger)aPieceNum {
    self = [super init];
    if (self) {
        self.pieceNum = aPieceNum;
        self.pieces = somePieces;
    }
    return self;
}

- (id)copyWithZone:(NSZone *)zone {
    // Immutable, so it's safe to return self
    return self;
}

- (NSString *)description {
    return [self toString];
}

#pragma mark -
#pragma mark Public methods

- (NSString *)getFront {
    if (self.pieceNum >= self.pieces.count) {
        return nil;
    }
    return [self.pieces objectAtIndex:self.pieceNum];
}

/**
 * @return The number of segments in this path
 */
- (NSUInteger)length {
    return self.pieces.count - self.pieceNum;
}

- (FPath *)popFront {
    NSInteger newPieceNum = self.pieceNum;
    if (newPieceNum < self.pieces.count) {
        newPieceNum++;
    }
    return [[FPath alloc] initWithPieces:self.pieces andPieceNum:newPieceNum];
}

- (NSString *)getBack {
    if (self.pieceNum < self.pieces.count) {
        return [self.pieces lastObject];
    } else {
        return nil;
    }
}

- (NSString *)toString {
    return [self toStringWithTrailingSlash:NO];
}

- (NSString *)toStringWithTrailingSlash {
    return [self toStringWithTrailingSlash:YES];
}

- (NSString *)toStringWithTrailingSlash:(BOOL)trailingSlash {
    NSMutableString *pathString = [[NSMutableString alloc] init];
    for (NSInteger i = self.pieceNum; i < self.pieces.count; i++) {
        [pathString appendString:@"/"];
        [pathString appendString:[self.pieces objectAtIndex:i]];
    }
    if ([pathString length] == 0) {
        return @"/";
    } else {
        if (trailingSlash) {
            [pathString appendString:@"/"];
        }
        return pathString;
    }
}

- (NSString *)wireFormat {
    if ([self isEmpty]) {
        return @"/";
    } else {
        NSMutableString *pathString = [[NSMutableString alloc] init];
        for (NSInteger i = self.pieceNum; i < self.pieces.count; i++) {
            if (i > self.pieceNum) {
                [pathString appendString:@"/"];
            }
            [pathString appendString:[self.pieces objectAtIndex:i]];
        }
        return pathString;
    }
}

- (FPath *)parent {
    if (self.pieceNum >= self.pieces.count) {
        return nil;
    } else {
        NSMutableArray *newPieces = [[NSMutableArray alloc] init];
        for (NSInteger i = self.pieceNum; i < self.pieces.count - 1; i++) {
            [newPieces addObject:[self.pieces objectAtIndex:i]];
        }
        return [[FPath alloc] initWithPieces:newPieces andPieceNum:0];
    }
}

- (FPath *)child:(FPath *)childPathObj {
    NSMutableArray *newPieces = [[NSMutableArray alloc] init];
    for (NSInteger i = self.pieceNum; i < self.pieces.count; i++) {
        [newPieces addObject:[self.pieces objectAtIndex:i]];
    }

    for (NSInteger i = childPathObj.pieceNum; i < childPathObj.pieces.count;
         i++) {
        [newPieces addObject:[childPathObj.pieces objectAtIndex:i]];
    }

    return [[FPath alloc] initWithPieces:newPieces andPieceNum:0];
}

- (FPath *)childFromString:(NSString *)childPath {
    NSMutableArray *newPieces = [[NSMutableArray alloc] init];
    for (NSInteger i = self.pieceNum; i < self.pieces.count; i++) {
        [newPieces addObject:[self.pieces objectAtIndex:i]];
    }

    NSArray *pathPieces = [childPath componentsSeparatedByString:@"/"];
    for (unsigned int i = 0; i < pathPieces.count; i++) {
        NSString *piece = [pathPieces objectAtIndex:i];
        if (piece.length > 0) {
            [newPieces addObject:piece];
        }
    }

    return [[FPath alloc] initWithPieces:newPieces andPieceNum:0];
}

/**
 * @return True if there are no segments in this path
 */
- (BOOL)isEmpty {
    return self.pieceNum >= self.pieces.count;
}

/**
 * @return Singleton to represent an empty path
 */
+ (FPath *)empty {
    static dispatch_once_t oneEmptyPath;
    static FPath *emptyPath;
    dispatch_once(&oneEmptyPath, ^{
      emptyPath = [[FPath alloc] initWith:@""];
    });
    return emptyPath;
}

- (BOOL)contains:(FPath *)other {
    if (self.length > other.length) {
        return NO;
    }

    NSInteger i = self.pieceNum;
    NSInteger j = other.pieceNum;
    while (i < self.pieces.count) {
        NSString *thisSeg = [self.pieces objectAtIndex:i];
        NSString *otherSeg = [other.pieces objectAtIndex:j];
        if (![thisSeg isEqualToString:otherSeg]) {
            return NO;
        }
        ++i;
        ++j;
    }
    return YES;
}

- (void)enumerateComponentsUsingBlock:(void (^)(NSString *, BOOL *))block {
    BOOL stop = NO;
    for (NSInteger i = self.pieceNum; !stop && i < self.pieces.count; i++) {
        block(self.pieces[i], &stop);
    }
}

- (NSComparisonResult)compare:(FPath *)other {
    NSInteger myCount = self.pieces.count;
    NSInteger otherCount = other.pieces.count;
    for (NSInteger i = self.pieceNum, j = other.pieceNum;
         i < myCount && j < otherCount; i++, j++) {
        NSComparisonResult comparison = [FUtilities compareKey:self.pieces[i]
                                                         toKey:other.pieces[j]];
        if (comparison != NSOrderedSame) {
            return comparison;
        }
    }
    if (self.length < other.length) {
        return NSOrderedAscending;
    } else if (other.length < self.length) {
        return NSOrderedDescending;
    } else {
        NSAssert(self.length == other.length,
                 @"Paths must be the same lengths");
        return NSOrderedSame;
    }
}

/**
 * @return YES if paths are the same
 */
- (BOOL)isEqual:(id)other {
    if (other == self) {
        return YES;
    }
    if (!other || ![other isKindOfClass:[self class]]) {
        return NO;
    }
    FPath *otherPath = (FPath *)other;
    if (self.length != otherPath.length) {
        return NO;
    }
    for (NSUInteger i = self.pieceNum, j = otherPath.pieceNum;
         i < self.pieces.count; i++, j++) {
        if (![self.pieces[i] isEqualToString:otherPath.pieces[j]]) {
            return NO;
        }
    }
    return YES;
}

- (NSUInteger)hash {
    NSUInteger hashCode = 0;
    for (NSInteger i = self.pieceNum; i < self.pieces.count; i++) {
        hashCode = hashCode * 37 + [self.pieces[i] hash];
    }
    return hashCode;
}

@end

File: FirebaseDatabaseSwiftCore/Sources/FPath.swift
//
//  File.swift
//  File
//
//  Created by Morten Bek Ditlevsen on 14/09/2021.
//

import Foundation

private let emptyPath = FPath(with: "")

@objc public class FPath: NSObject, NSCopying {
    let pieceNum: Int
    let pieces: [String]

    @objc public class func relativePath(from outer: FPath, to inner: FPath) -> FPath {
        guard let outerFront = outer.getFront() else {
            return inner
        }
        let innerFront = inner.getFront()
        if outerFront == innerFront {
            return relativePath(from: outer.popFront(), to: inner.popFront())
        } else {
            fatalError()
//            @throw [[NSException alloc]
//                initWithName:@"FirebaseDatabaseInternalError"
//                      reason:[NSString
//                                 stringWithFormat:
//                                     @"innerPath (%@) is not within outerPath (%@)",
//                                     inner, outer]
//                    userInfo:nil];
        }
    }

    @objc public class var empty: FPath { emptyPath }

    @objc public class func path(string: String) -> FPath {
        FPath(with: string)
    }

    @objc public init(with path: String) {
        let pathPieces = path.components(separatedBy: "/")
        self.pieces = pathPieces.filter {
            !$0.isEmpty
        }
        self.pieceNum = 0
    }

    @objc public init(pieces: [String], andPieceNum pieceNum: Int) {
        self.pieces = pieces
        self.pieceNum = pieceNum
    }

    @objc public func copy(with zone: NSZone? = nil) -> Any {
        // Immutable, so it's safe to return self
        return self
    }

    @objc public func enumerateComponents(usingBlock block: @escaping (_ key: String, _ stop: UnsafeMutablePointer<ObjCBool>) -> Void) {
        var stop: ObjCBool = false
        for piece in pieces[pieceNum...] {
            withUnsafeMutablePointer(to: &stop) { pointer in
                block(piece, pointer)
            }
            if stop.boolValue { break }
        }
    }

    @objc public func getFront() -> String? {
        guard pieceNum < pieces.count else {
            return nil
        }
        return pieces[pieceNum]
    }

    @objc public func length() -> Int {
        pieces.count - pieceNum
    }

    @objc public func popFront() -> FPath {
        var newPieceNum = pieceNum
        if newPieceNum < pieces.count {
            newPieceNum += 1
        }
        return FPath(pieces: pieces, andPieceNum: newPieceNum)
    }

    @objc public func getBack() -> String? {
        pieces.last
    }

    @objc public func toString() -> String {
        toString(withTrailingSlash: false)
    }

    @objc public override var description: String {
        toString()
    }

    @objc public func toStringWithTrailingSlash() -> String {
        toString(withTrailingSlash: true)
    }

    private func toString(withTrailingSlash trailingSlash: Bool) -> String {
        var pathString = ""
        for piece in pieces[pieceNum...] {
            pathString += "/\(piece)"
        }

        if pathString.isEmpty {
            return "/"
        } else {
            if trailingSlash {
                return pathString + "/"
            }
            return pathString
        }
    }

    @objc public func wireFormat() -> String {
        isEmpty ? "/" : pieces[pieceNum...].joined(separator: "/")
    }

    @objc public func parent() -> FPath? {
        guard pieceNum < pieces.count else {
            return nil
        }
        
        return FPath(pieces: Array(pieces[pieceNum..<(pieces.count - 1)]), andPieceNum: 0)
    }

    @objc public func child(_ childPathObj: FPath) -> FPath {
        var newPieces = Array(pieces[pieceNum...])
        newPieces.append(contentsOf: childPathObj.pieces[childPathObj.pieceNum...])
        return FPath(pieces: newPieces, andPieceNum: 0)
    }

    @objc public func child(fromString childPath: String) -> FPath {
        var newPieces = Array(pieces[pieceNum...])

        let pathPieces = childPath.components(separatedBy: "/")
        newPieces.append(contentsOf: pathPieces.filter {
            !$0.isEmpty
        })

        return FPath(pieces: newPieces, andPieceNum: 0)
    }

    @objc public var isEmpty: Bool {
        pieceNum >= pieces.count
    }

    @objc public func contains(_ other: FPath) -> Bool {
        guard self.length() <= other.length() else {
            return false
        }

        for (a, b) in zip(pieces[pieceNum...], other.pieces[other.pieceNum...]) {
            if a != b {
                return false
            }
        }
        return true
    }

    @objc public func compare(_ other: FPath) -> ComparisonResult {
        for (a, b) in zip(pieces[pieceNum...], other.pieces[other.pieceNum...]) {
            let comparison = FUtilitiesSwift.compareKey(a, b)
            if comparison != .orderedSame {
                return comparison;
            }
        }
        if (self.length() < other.length()) {
            return .orderedAscending
        } else if other.length() < self.length() {
            return .orderedDescending
        } else {
            assert(self.length() == other.length(),
                     "Paths must be the same lengths")
            return .orderedSame
        }

    }

    @objc override public func isEqual(_ other: Any?) -> Bool {
        guard let other = other as? FPath else { return false }
        if other === self {
            return true
        }
        if length() != other.length() {
            return false
        }
        for (a, b) in zip(pieces[pieceNum...], other.pieces[other.pieceNum...]) {
            if a != b { return false }
        }
        return true
    }

    @objc override public var hash: Int {
        var hasher = Hasher()
        for piece in pieces[pieceNum...] {
            piece.hash(into: &hasher)
        }
        return hasher.finalize()
    }
}
